### PR TITLE
docs: initial look at vms and ephemeral runners

### PR DESCRIPTION
**Pull request by weikequ**
Signed-off-by: Weike Qu <weikequ@amazon.com>

*Issue #, if available:*

*Description of changes:*

*Testing done:*

- [ ] I've reviewed the guidance in CONTRIBUTING.md

#### License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/ephemeral-macos-runners.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+### Option 1 - Use virtualization on EC2 instances & ephemeral runners - **Recommended**
+
+Pros
+- Solves all the issues above
+- Can have 1 dedicated host run up to 2 VMs/runners, effectively halving our infrastructure costs
+- Could be an interesting contribution back to the community
+
+Cons
+- Needs engineering time
```
Comment (davidhsingyuchen) [inline docs/ephemeral-macos-runners.md:48]: Is there a rough plan and/or estimate for this?

File: docs/ephemeral-macos-runners.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+- Will only be needed until we can get amd64/arm64 GitHub hosted runners (and get them to work properly) - this is a big variable 
+- May have to purchase/depend on additional software (eg. [Veertu Anka Build](https://aws.amazon.com/marketplace/pp/prodview-rp7kfyl56arbe?sr=0-1&ref_=beagle&applicationId=AWSMPContessa))
```
Comment (davidhsingyuchen) [inline docs/ephemeral-macos-runners.md:50]: Would it offset the cost saved on EC2?

File: docs/ephemeral-macos-runners.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    - Used for environment dependent tasks such as release builds & e2e tests.
+- EC2 hosted arm (Apple silicon) macOS runners
+    - Used for environment dependent tasks such as release builds & e2e tests.
+
+In the **long term**, all infrastructure should be run using GitHub infrastructure, but currently, key issues exist that preclude that possibility.
+
+- x86 macOS runners
+    - Currently available on GitHub
+    - Speed - As of v0.1.0, vm init can take up to 10 minutes to execute. This makes running many e2e tests and config tests very impractical.
+    - Bug with `finch pull` that makes each `finch pull` action take 10x as long to complete.
```
Comment (monirul) [inline docs/ephemeral-macos-runners.md:19]: Is this related with runner? or finch bug? If finch bug then may be we can add the issue link here

**Comment by pendo324** [comment]
Has [this](https://github.com/machulav/ec2-github-runner) GitHub action ever been evaluated? Looks like it does some of what is described in the doc

File: docs/ephemeral-macos-runners.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+### Option 2 - Do job level clean-up
+
+Pros
+- Easy to do
+- Still better than doing nothing
+
+Cons
+- Still require our current # of hosts
+- Not scalable - have to add clean up step to every single workflow
```
Comment (davidhsingyuchen) [inline docs/ephemeral-macos-runners.md:60]: For `finch`, it's [probably manageable](https://github.com/runfinch/finch/blob/1da659f228458f69d4adb67b6d1695477b1ae4d4/.github/workflows/ci.yaml#L87-L97). Wondering how painful it is for finch-core.

File: docs/ephemeral-macos-runners.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+- EC2 hosted x86 (intel) macOS runners
+    - Used for environment dependent tasks such as release builds & e2e tests.
+- EC2 hosted arm (Apple silicon) macOS runners
+    - Used for environment dependent tasks such as release builds & e2e tests.
+
+In the **long term**, all infrastructure should be run using GitHub infrastructure, but currently, key issues exist that preclude that possibility.
+
+- x86 macOS runners
+    - Currently available on GitHub
+    - Speed - As of v0.1.0, vm init can take up to 10 minutes to execute. This makes running many e2e tests and config tests very impractical.
```
Comment (ningziwen) [inline docs/ephemeral-macos-runners.md:18]: Maybe make both vm init and finch pull as one generic one like "the commands related to network transferring into VM", then for example "vm init", "finch pull".

File: docs/ephemeral-macos-runners.md (hunk 3 above)
Comment (weikequ) [reply docs/ephemeral-macos-runners.md:19]: It's runner related. In our testing before, it keeps happens exclusively on GitHub hosted macOS runners, not EC2 instances nor our local machines.

Comment (davidhsingyuchen) [reply docs/ephemeral-macos-runners.md:19]: Maybe we can also say 10x of what (i.e., `finch pull` in other environment, or `pull` by other container development tool in the same environment, etc.).

File: docs/ephemeral-macos-runners.md (hunk 5 above)
Comment (weikequ) [reply docs/ephemeral-macos-runners.md:18]: No, I disagree. I believe they are fundamentally different issues. One (init) is caused by the processing speed of the machine, one (pull) is caused by an unknown issue.

File: docs/ephemeral-macos-runners.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    - With performant runners that run on the bare metal dedicated hosts, we need to be wary of controlling the lifecycle of the runner and its created subprocesses for security risks.
+    - Our current runners do not guarantee 100% of the subprocesses and created files are eliminated upon termination of the run.
+- Efficiency
+    - Our current system has a 1:1 relationship with the # of dedicated hosts mapped to the # of runners. Given the powerful multi-core nature of the EC2 dedicated hosts, this is an extreme underutilization of resources and actively costs us money.
+- Stability
+    - Each run does not run on a blank, fresh image. This means that there could be leftover processes/files/settings/etc that may impact future runs. This creates situations that are less reproducible and leads to more engineering time spent troubleshooting very specific issues with the infrastructure rather than the code.
+
+## Proposed Solutions
+
+### Option 1 - Use virtualization on EC2 instances & ephemeral runners - **Recommended**
```
Comment (ningziwen) [inline docs/ephemeral-macos-runners.md:40]: Could you elaborate the feasibility?

File: docs/ephemeral-macos-runners.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+Pros
+- Solves all the issues above
+- Can have 1 dedicated host run up to 2 VMs/runners, effectively halving our infrastructure costs
+- Could be an interesting contribution back to the community
+
+Cons
```
Comment (monirul) [inline docs/ephemeral-macos-runners.md:47]: Did we check if there is any performance penalty? since we are running two VMs in one host.

File: docs/ephemeral-macos-runners.md (hunk 5 above)
Comment (ningziwen) [reply docs/ephemeral-macos-runners.md:18]: Yeah my intuition is they may have same root cause. If we are not sure about the root causes, I think it is ok to leave as it is now.
